graph_test.cpp

```diff
@@ -366,4 +382,62 @@ TEST(GraphTest, shouldNoticeWhenStopped) {
     EXPECT_FALSE(graph.isStarted());
 }
 
+TEST(GraphTest, shouldAddAndRemoveNodesWhileRunning) {
+    Graph graph;
+    auto producer = graph.newNode<ThreadedIntProducer>("producer");
+
+    std::vector<std::shared_ptr<ThreadedIntConsumer>> consumers;
+
+    std::mt19937 gen(42);
+    std::uniform_int_distribution<> dice6(1, 6);
+    int nextId = 0;
+    size_t maxConsumers = 0;
+    size_t totalConsumed = 0;
+
+    EXPECT_TRUE(graph.start());
+    for (Timestamp endTime = Timestamp::now() + Duration::seconds(5); Timestamp::now() < endTime;
+         Duration::milliSeconds(3).sleep()) {
+        switch (dice6(gen)) {
+            case 1: {
+                // create a new consumer
+                std::stringstream ss;
+                ss << "consumer_" << nextId++;
+                auto consumer =
+                    graph.newNode<ThreadedIntConsumer>(ss.str(), Duration::milliSeconds(2));
+                consumers.push_back(consumer);
+                EXPECT_TRUE(graph.connect(producer, "out", consumer, "in"));
+                break;
+            }
+            case 2:
+                if (!consumers.empty()) {
+                    // randomly kill a consumer
+                    auto it = consumers.begin() + std::uniform_int_distribution<long>(
+                                                      0, long(consumers.size() - 1))(gen);
+                    auto pin = (*it)->inputPin(0);
+                    pin->disconnect();
+                    totalConsumed += (*it)->consumed();
+                    graph.removeNode((*it)->name());
+                    consumers.erase(it);
```

Comment:
if you want to make sure the node was removed and that no new messages where sent to it, it could be moved to another list and 'frozen', so that it aborts if it receives a message. Maybe it's a bit extreme, but may make sense as an extra test

Reply:
Good idea. But I also want to test what happen when destroying during graph execution, so I randomly pick a behavior, having 50% stopped now and 50% kept running.